### PR TITLE
Tokenizers API developments

### PR DESCRIPTION
**Pull request by thomwolf**
Some developments on the tokenizer's API to make it more flexible and easy to use for dynamic/uniform size batching.

- add `return_lengths` methods to both slow and fast tokenizer to have a column with input lengths (useful to sort them afterward)
- make `tokenizer.pad()` accept a list of dicts so it can be used as a `collate_fn` in a PyTorch data loader.
- safety check that all the `kwargs` inputs to encoding methods are recognized (avoid silent errors). raise a warning if not, not an error for now (reduce breaking change behaviors).
- standardized arguments of the encoding methods in unified names (`return_attention_masks` => `return_attention_mask` and `return_special_tokens_masks` => `return_special_tokens_mask`). This is a slight breaking change but even the maintainers of the lib were actually making the confusion as well (which usually resulted in silent error hidden behind the above mentioned generic use of kwargs in encoding errors).
- use the `tokenizers` library `AddedToken` class to control more finely how added (and special) tokens are tokenized (do we tokenize them inside words, do we strip left and/or right white spaces around them). This class gives a simpler and cleaner behavior for `RobertaTokenizer`.
- change the way special tokens can be added to the tokenizers. We now have more separated paths:

   * giving a string at tokenizer initialization or setting an attribute (e.g. `tokenizer.mask_token`) update the relevant attribute in the tokenizer class but doesn't add tokens in the vocabulary. We do that because the initialization of the special tokens is usually done before the backend vocabulary is setup. This automatic and non explicit addition of tokens can also be error-prone and a source of silent-errors.
   * the recommended way to add a token and use it as a special token is to use `add_special_tokens` which will store the attribute AND add the token to the vocabulary if needed and which returns the number of added tokens
   * a new method `tokenizer.sanitize_special_tokens()` can be used to make sure all the special tokens are in the vocabulary and to automatically add them if it's not the case.

This PR should in particular fix cleanly the Roberta specific behavior discussed in:
- https://github.com/huggingface/transformers/pull/2778
- https://github.com/huggingface/transformers/issues/3788

Roberta now behave like GPT2, i.e. without a prefix token. It can still be used in the masked-fill pipeline since with can control the behavior of the mask token with the parameters of `AddedToken` and in particular keep a space after the mask token as required to use the pipeline correctly.

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -204,24 +259,63 @@ def tokenize(self, text: TextInput, **kwargs):
...
 
         def split_on_token(tok, text):
             result = []
+            tok_extended = all_special_tokens_extended.get(tok, None)
             split_text = text.split(tok)
+            full_word = ""
             for i, sub_text in enumerate(split_text):
-                sub_text = sub_text.rstrip()
+                # AddedToken can control whitespace stripping around them.
+                # We use them for GPT2 and Roberta to have differential behavior depending on the special token
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils.py:286]: Differential? I'd say different but it's just me

Comment (LysandreJik) [reply src/transformers/tokenization_utils.py:286]: Offering a suggestion for easy commit in the case you meant different
```suggestion
                # We use them for GPT2 and Roberta to have different behavior depending on the special token
```

File: tests/test_tokenization_roberta.py [hunk 2]
```diff
@@ -80,12 +80,12 @@ def get_input_output_texts(self, tokenizer):
     def test_full_tokenizer(self):
         tokenizer = RobertaTokenizer(self.vocab_file, self.merges_file, **self.special_tokens_map)
         text = "lower newer"
-        bpe_tokens = ["\u0120low", "er", "\u0120", "n", "e", "w", "er"]
-        tokens = tokenizer.tokenize(text, add_prefix_space=True)
+        bpe_tokens = ["l", "o", "w", "er", "\u0120", "n", "e", "w", "er"]
```
Comment (sshleifer) [inline tests/test_tokenization_roberta.py:83]: does this show that this PR changes the default tokenization?

File: tests/test_tokenization_fast.py [hunk 3]
```diff
@@ -63,6 +63,21 @@ def test_all_tokenizers(self):
                         self.fast_align_python(tokenizer_r, tokenizer_p, tok_case, pretrained_name)
                         self.fast_only(tokenizer_r)
 
+    def test_pretokenized_tokenizers(self):
+        for tok_case in self.TOKENIZERS_CLASSES:
```
Comment (sshleifer) [inline tests/test_tokenization_fast.py:67]: `tok_cls`

File: tests/test_tokenization_roberta.py (hunk 2 above)
Comment (sshleifer) [reply tests/test_tokenization_roberta.py:83]: Do we still want test coverage for add_prefix_space=True?

**Review by LysandreJik — APPROVED** [review]
Very cool, looks great to me!

File: tests/test_tokenization_roberta.py (hunk 2 above)
Comment (thomwolf) [reply tests/test_tokenization_roberta.py:83]: Dynamically adding `add_prefix_space=True` at tokenization time is not possible in fast tokenizers anymore and I think it should be deprecated in the future in slow tokenizers as well. I'll add a DeprecationWarning.

File: src/transformers/tokenization_utils.py [hunk 4]
```diff
@@ -124,7 +177,7 @@ def __len__(self):
         """ Size of the full vocabulary with the added tokens """
         return self.vocab_size + len(self.added_tokens_encoder)
 
-    def add_tokens(self, new_tokens: Union[str, List[str]]) -> int:
+    def add_tokens(self, new_tokens: Union[str, List[str]], special_token=False) -> int:
```
Comment (n1t0) [inline src/transformers/tokenization_utils.py:180]: What's the rationale for adding this? It doesn't seem to be used below.

File: src/transformers/tokenization_utils_fast.py [hunk 5]
```diff
@@ -235,16 +233,12 @@ def add_tokens(self, new_tokens: List[Union[str, AddedTokenFast]]) -> int:
...
-        # TODO This should be done in tokenizers to be really clean.
-        # Removing for now
-        # tokens = []
-        # for token in new_tokens:
-        #     if self.init_kwargs.get("do_lower_case", False) and token not in self.all_special_tokens:
-        #         token = token.lower()
-        #     if token not in tokens:
-        #         tokens.append(token)
+
+        if special_token:
```
Comment (n1t0) [inline src/transformers/tokenization_utils_fast.py:239]: Ok, got it

File: src/transformers/tokenization_utils_fast.py [hunk 6]
```diff
@@ -235,16 +233,12 @@ def add_tokens(self, new_tokens: List[Union[str, AddedTokenFast]]) -> int:
             print('We have added', num_added_toks, 'tokens')
             model.resize_token_embeddings(len(tokenizer))  # Notice: resize_token_embeddings expect to receive the full size of the new vocabulary, i.e. the length of the tokenizer.
         """
-        if isinstance(new_tokens, str):
+        if not isinstance(new_tokens, (list, tuple)):
```
Comment (n1t0) [inline src/transformers/tokenization_utils_fast.py:236]: I'm not sure it works with a tuple actually, but I guess it should indeed.

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (thomwolf) [reply src/transformers/tokenization_utils.py:180]: It's used for the fast tokenizers here: https://github.com/huggingface/transformers/blob/master/src/transformers/tokenization_utils_fast.py#L209-L242

Comment (thomwolf) [reply src/transformers/tokenization_utils.py:180]: Kept the same signature in both fast and slow tokenizers
